src/backend/dsg_updater.cpp

```diff
@@ -134,30 +144,62 @@ void DsgUpdater::callUpdateFunctions(size_t timestamp_ns, UpdateInfo::ConstPtr i
 
     functor->call(*source_graph_, *target_dsg_, info);
     if (hooks.find_merges && enable_merging) {
+      auto& tracker = merge_tracker.getMergeGroup(name);
+
       // TODO(nathan) handle given merges
       const auto merges = hooks.find_merges(*source_graph_, info);
       const auto applied =
-          merge_tracker.applyMerges(*source_graph_, merges, *target_dsg_, hooks.merge);
+          tracker.applyMerges(*source_graph_, merges, *target_dsg_, hooks.merge);
       VLOG(1) << "[Backend: " << name << "] Found " << merges.size()
               << " merges (applied " << applied << ")";
 
       if (config.enable_exhaustive_merging) {
         size_t merge_iter = 0;
         size_t num_applied = 0;
         do {
+          if (name == "surface_places") {
+            continue;
+          }
+
           const auto new_merges = hooks.find_merges(*target_dsg_->graph, info);
-          num_applied = merge_tracker.applyMerges(
-              *target_dsg_->graph, new_merges, *target_dsg_, hooks.merge);
+          num_applied = tracker.applyMerges(*source_graph_,  // Fixes bug?
+                                            new_merges,
+                                            *target_dsg_,
+                                            hooks.merge);
           VLOG(1) << "[Backend: " << name << "] Found " << new_merges.size()
                   << " merges at pass " << merge_iter << " (" << num_applied
                   << " applied)";
           ++merge_iter;
         } while (num_applied > 0);
       }
+      if (info->loop_closure_detected && hooks.merge) {
+        LOG(WARNING) << "Updating all merge attributes for " << name;
+        tracker.print();
+        tracker.updateAllMergeAttributes(
+            *source_graph_, *target_dsg_->graph, hooks.merge);
+      }
     }
-  }
 
-  launchCallbacks(cleanup_hooks, info, target_dsg_.get());
+    merge_tracker.print();
+
+    // NOTE: the followings fails with hundreds of lines of errors about templates and
+    // threads:
+    // launchCallbacks(cleanup_hooks, info, *source_graph_, target_dsg_.get());
```

Comment:
Anything passed by reference to a std::thread constructor gets copied, the arg needs to be raw / shared pointer for this to work